src/cata_ui.cpp

```diff
@@ -0,0 +1,442 @@
+
+    //erase the top 3 rows
+    for( unsigned int y = 0; y < 3; y++ ) {
+        for( unsigned int x = 0; x < get_rect().size_x; x++ ){
+            mvwputch( win, y, x, border_color, ' ' );
+        }
+    }
+
+    for ( unsigned int i = 0; i < get_rect().size_x; i++ ) {
+        mvwputch( win, 2, i, border_color, LINE_OXOX );
+    }
+
+    mvwputch( win, 2,  0, border_color, LINE_OXXO ); // |^
+    mvwputch( win, 2, get_rect().size_x - 1, border_color, LINE_OOXX ); // ^|
+
+    int x_offset = 1; // leave space for selection bracket
+    for( unsigned int i = 0; i < tabs.size(); i++ ) {
+        x_offset += draw_tab( win, x_offset, tabs[i].first, tab_index == i ) + 2;
+    }
+}
+
+void tabbed_window::next_tab()
+{
+    if( tabs.size() > 1 ) {
+        tabs[tab_index].second->set_visible( false );
+        if( tab_index == tabs.size() - 1 ){
+            tab_index = 0;
+        } else {
+            tab_index++;
+        }
+        tabs[tab_index].second->set_visible( true );
+
+        on_scroll();
+    }
+}
+
+void tabbed_window::previous_tab()
+{
+    if( tabs.size() > 1 ) {
+        tabs[tab_index].second->set_visible( false );
+        if( tab_index == 0 ){
+            tab_index = tabs.size() - 1;
+        } else {
+            tab_index--;
+        }
+        tabs[tab_index].second->set_visible( true );
+
+        on_scroll();
+    }
+}
+
+std::string tabbed_window::current_tab() const
+{
+    if( tabs.empty() ) {
+        return "";
+    }
+    return tabs[tab_index].first;
+}
+
+input_broadcaster::listener tabbed_window::default_action_handler()
+{
+    return {
+        { "PREV_TAB", "NEXT_TAB" },
+        [this]( const std::string &action ) {
+            if( action == "PREV_TAB" ) {
+                previous_tab();
+            } else if( action == "NEXT_TAB" ) {
+                next_tab();
+            }
+        }
+    };
+}
+
+ui_horizontal_list::ui_horizontal_list( int x, int y, ui_anchor anchor ) : ui_element( 0, 1, x, y, anchor )
+{
+}
+
+void ui_horizontal_list::draw( WINDOW *win )
+{
+    unsigned int x_offset = 1 + get_ax();
+
+    for( unsigned int i = 0; i < text.size(); i++ ) {
+        x_offset += draw_subtab( win, get_ax() + x_offset, get_ay(), text[i], i == scroll ) + 2;
+    }
+}
+
+void ui_horizontal_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+    size_t text_length = 1;
+
+    for( const auto &str : text ) {
+        text_length += utf8_width( str ) + 1;
+    }
+
+    set_rect( ui_rect( text_length, get_rect().size_y, get_rect().x, get_rect().y ) );
+}
+
+void ui_horizontal_list::scroll_left()
+{
+    if( text.size() > 1) {
+        scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+        on_scroll();
+    }
+}
+
+void ui_horizontal_list::scroll_right()
+{
+    if( text.size() > 1 ) {
+        scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+        on_scroll();
+    }
+}
+
+const std::string &ui_horizontal_list::current() const
+{
+    return text[scroll];
```

Comment:
This function has a slightly different (and therefor surprising) behavior compared to `tabbed_window::current_tab`. The later works fine even if the tab list is empty (returns an empty string), the former will drift into undefined behavior as it accesses the string vector anyway.

Also: why is this named just `current`, but the other function is named `current_tab`?